File: skills/onboard/scripts/register-plugin.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');

const PLUGIN_NAME = 'second-brain';
const MARKETPLACE_NAME = 'second-brain';
const ENABLED_KEY = `${PLUGIN_NAME}@${MARKETPLACE_NAME}`;

function die(msg) {
  process.stderr.write(`error: ${msg}\n`);
  process.exit(1);
}

function parseArgs(argv) {
  const opts = { scope: null, vault: null };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--scope') opts.scope = argv[++i];
    else if (a === '--vault') opts.vault = argv[++i];
    else die(`unknown argument: ${a}`);
  }
  if (opts.scope !== 'project' && opts.scope !== 'user') {
    die(`--scope must be 'project' or 'user'`);
  }
  if (opts.scope === 'project' && !opts.vault) {
    die(`--vault is required when --scope=project`);
  }
  return opts;
}

function findPluginRoot(start) {
  let dir = start;
  while (true) {
    if (fs.existsSync(path.join(dir, '.claude-plugin', 'plugin.json'))) return dir;
    const parent = path.dirname(dir);
    if (parent === dir) die(`could not find .claude-plugin/plugin.json walking up from ${start}`);
    dir = parent;
  }
}

function targetSettingsPath(scope, vault) {
  if (scope === 'project') {
    return path.join(path.resolve(vault), '.claude', 'settings.json');
  }
  return path.join(process.env.HOME || os.homedir(), '.claude', 'settings.json');
}

function loadExisting(filePath) {
  if (!fs.existsSync(filePath)) return {};
  const raw = fs.readFileSync(filePath, 'utf8');
  let data;
  try { data = JSON.parse(raw); }
  catch (err) { die(`${filePath} is not valid JSON: ${err.message}`); }
  if (data === null || typeof data !== 'object' || Array.isArray(data)) {
    die(`${filePath} top-level value must be a JSON object`);
  }
  return data;
}

function sortKeys(value) {
  if (Array.isArray(value)) return value.map(sortKeys);
  if (value !== null && typeof value === 'object') {
    return Object.keys(value).sort().reduce((acc, k) => {
      acc[k] = sortKeys(value[k]);
      return acc;
    }, {});
  }
  return value;
}

function merge(settings, pluginPath) {
  if (!settings.extraKnownMarketplaces || typeof settings.extraKnownMarketplaces !== 'object') {
    settings.extraKnownMarketplaces = {};
  }
  settings.extraKnownMarketplaces[MARKETPLACE_NAME] = {
    source: { source: 'directory', path: pluginPath },
  };
  if (!settings.enabledPlugins || typeof settings.enabledPlugins !== 'object') {
    settings.enabledPlugins = {};
  }
  settings.enabledPlugins[ENABLED_KEY] = true;
  return settings;
}

function writeAtomically(filePath, data) {
  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  const tmp = `${filePath}.tmp`;
  const sorted = sortKeys(data);
  fs.writeFileSync(tmp, JSON.stringify(sorted, null, 2) + '\n');
  fs.renameSync(tmp, filePath);
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  const pluginRoot = findPluginRoot(path.dirname(fs.realpathSync(__filename)));
  const target = targetSettingsPath(args.scope, args.vault);
  const existing = loadExisting(target);
  const merged = merge(existing, pluginRoot);
  writeAtomically(target, merged);
  process.stderr.write(`wrote ${target}\n`);
}

main();

File: skills/onboard/scripts/register-plugin.py
#!/usr/bin/env python3
"""Register the second-brain plugin into a Claude Code settings.json.

The script is self-locating: it walks up from its own __file__ to find the
plugin root (the directory containing .claude-plugin/plugin.json) and uses
that absolute path as the `directory` source.

Usage:
    register-plugin.py --scope project --vault <abs-path-to-vault>
    register-plugin.py --scope user

For --scope project, writes/merges <vault>/.claude/settings.json.
For --scope user,    writes/merges $HOME/.claude/settings.json.

Existing values for `extraKnownMarketplaces.second-brain` and
`enabledPlugins["second-brain@second-brain"]` are overwritten. Every other
key in the file is preserved. Malformed JSON in the target file is a fatal
error — the script exits non-zero and does NOT touch the file.
"""

import argparse
import json
import os
import sys
from pathlib import Path

PLUGIN_NAME = "second-brain"
MARKETPLACE_NAME = "second-brain"
ENABLED_KEY = f"{PLUGIN_NAME}@{MARKETPLACE_NAME}"


def find_plugin_root(start: Path) -> Path:
    for candidate in [start, *start.parents]:
        if (candidate / ".claude-plugin" / "plugin.json").is_file():
            return candidate
    sys.exit(
        f"error: could not find .claude-plugin/plugin.json walking up from {start}"
    )


def target_settings_path(scope: str, vault: str | None) -> Path:
    if scope == "project":
        if not vault:
            sys.exit("error: --vault is required when --scope=project")
        return Path(vault).expanduser().resolve() / ".claude" / "settings.json"
    if scope == "user":
        return Path(os.environ.get("HOME", str(Path.home()))) / ".claude" / "settings.json"
    sys.exit(f"error: unknown scope {scope!r}")


def load_existing(path: Path) -> dict:
    if not path.exists():
        return {}
    try:
        with path.open("r") as f:
            data = json.load(f)
    except json.JSONDecodeError as exc:
        sys.exit(f"error: {path} is not valid JSON: {exc}")
    if not isinstance(data, dict):
        sys.exit(f"error: {path} top-level value must be a JSON object")
    return data


def merge(settings: dict, plugin_path: Path) -> dict:
    mkts = settings.setdefault("extraKnownMarketplaces", {})
    mkts[MARKETPLACE_NAME] = {
        "source": {"source": "directory", "path": str(plugin_path)}
    }
    enabled = settings.setdefault("enabledPlugins", {})
    enabled[ENABLED_KEY] = True
    return settings


def write_atomically(path: Path, data: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    with tmp.open("w") as f:
        json.dump(data, f, indent=2, sort_keys=True)
        f.write("\n")
    tmp.replace(path)


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--scope", choices=["project", "user"], required=True)
    parser.add_argument("--vault", help="vault root (required when --scope=project)")
    args = parser.parse_args()

    plugin_root = find_plugin_root(Path(__file__).resolve())
    target = target_settings_path(args.scope, args.vault)
    settings = load_existing(target)
    merged = merge(settings, plugin_root)
    write_atomically(target, merged)
    print(f"wrote {target}", file=sys.stderr)


if __name__ == "__main__":
    main()
